src/hotspot/share/opto/parse1.cpp

```diff
@@ -1118,11 +1119,26 @@ void Parse::do_exits() {
       JVMState* ex_jvms = caller->clone_shallow(C);
       ex_jvms->bind_map(kit.clone_map());
       ex_jvms->set_bci(   InvocationEntryBci);
-      kit.set_jvms(ex_jvms);
+
       if (do_synch) {
 #ifdef C2_PATCH
         if (ObjectMonitorMode::legacy()) {
 #endif
+        // Ensure that the jvms object has a method pointer inorder to
+        // avoid asserts.
+        if (ObjectMonitorMode::java()) {
+          if (!ex_jvms->has_method()){
+            ex_jvms->set_method(method());
```

Comment:
We probably don't need this if we find out why the caller state doesn't have a method.  I'll see if I can find the problem.